src/storage/buffer_manager.cpp

```diff
@@ -339,32 +384,41 @@ BufferHandle BufferManager::Pin(shared_ptr<BlockHandle> &handle) {
 		}
 		required_memory = handle->memory_usage;
 	}
-	D_ASSERT(required_memory >= Storage::BLOCK_SIZE);
 	// evict blocks until we have space for the current block
 	unique_ptr<FileBuffer> reusable_buffer;
-	if (!EvictBlocks(required_memory, maximum_memory, &reusable_buffer)) {
-		throw OutOfMemoryException("failed to pin block of size %lld%s", required_memory, InMemoryWarning());
-	}
+	auto reservation = EvictBlocksOrThrow(required_memory, maximum_memory, &reusable_buffer,
+	                                      "failed to pin block of size %lld%s", required_memory);
 	// lock the handle again and repeat the check (in case anybody loaded in the mean time)
 	lock_guard<mutex> lock(handle->lock);
 	// check if the block is already loaded
 	if (handle->state == BlockState::BLOCK_LOADED) {
 		// the block is loaded, increment the reader count and return a pointer to the handle
 		handle->readers++;
-		D_ASSERT(current_memory >= required_memory);
-		current_memory -= required_memory;
+		reservation.Resize(current_memory, 0);
 		return handle->Load(handle);
 	}
 	// now we can actually load the current block
 	D_ASSERT(handle->readers == 0);
 	handle->readers = 1;
-	return handle->Load(handle, move(reusable_buffer));
+	auto buf = handle->Load(handle, move(reusable_buffer));
+	handle->memory_charge = move(reservation);
+	// In the case of a variable sized block, the buffer may be smaller than a full block.
+	int64_t delta = handle->buffer->AllocSize() - handle->memory_usage;
+	if (delta) {
+		D_ASSERT(delta < 0);
+		handle->memory_usage += delta;
+		handle->memory_charge.Resize(current_memory, handle->memory_usage);
+	}
+	return buf;
 }
 
 void BufferManager::AddToEvictionQueue(shared_ptr<BlockHandle> &handle) {
 	D_ASSERT(handle->readers == 0);
 	handle->eviction_timestamp++;
-	PurgeQueue();
+	// After each 1024 insertions, run through the queue and purge.
+	if ((++queue_insertions & 0x3ff) == 0) {
```

Comment:
Can we rewrite this to `& (1024 - 1)` for more clarity, and perhaps place `1024` into a separate constexpr?